File: editors/cfd-editor/src-tauri/src/commands/language.rs
//! 语言文档同步、诊断、补全与格式化命令。

use cfd_editor_core::EditorHost;
use tauri::State;

use super::run_host_command;
use crate::editor::{
    EditorError, FunctionDocumentState, LanguageCompletion, LanguageDocumentState,
    LanguageFormattingResult, LanguagePosition, ProjectBootstrap,
};
use coflow_runtime::FlatDiagnostic;
#[tauri::command]
pub(crate) async fn read_source_text(
    session_id: u32,
    file_path: String,
    host: State<'_, EditorHost>,
) -> Result<String, EditorError> {
    run_host_command(host, move |host| {
        host.sessions().read_source_text(session_id, &file_path)
    })
    .await
}

#[tauri::command]
pub(crate) async fn highlight_source_snapshot(
    session_id: u32,
    file_path: String,
    source: String,
    host: State<'_, EditorHost>,
) -> Result<LanguageDocumentState, EditorError> {
    run_host_command(host, move |host| {
        host.sessions()
            .highlight_source_snapshot(session_id, &file_path, &source)
    })
    .await
}

#[tauri::command]
pub(crate) async fn sync_language_document(
    session_id: u32,
    file_path: String,
    source: String,
    version: i64,
    host: State<'_, EditorHost>,
) -> Result<LanguageDocumentState, EditorError> {
    run_host_command(host, move |host| {
        host.sessions()
            .sync_language_document(session_id, &file_path, &source, version)
    })
    .await
}

#[tauri::command]
pub(crate) async fn validate_source_text(
    session_id: u32,
    file_path: String,
    source: String,
    host: State<'_, EditorHost>,
) -> Result<Vec<FlatDiagnostic>, EditorError> {
    run_host_command(host, move |host| {
        host.sessions()
            .validate_source_text(session_id, &file_path, &source)
    })
    .await
}

#[tauri::command]
pub(crate) async fn complete_language_document(
    session_id: u32,
    file_path: String,
    source: String,
    version: i64,
    position: LanguagePosition,
    host: State<'_, EditorHost>,
) -> Result<Vec<LanguageCompletion>, EditorError> {
    run_host_command(host, move |host| {
        host.sessions()
            .complete_language_document(session_id, &file_path, &source, version, &position)
    })
    .await
}

#[tauri::command]
pub(crate) async fn format_language_document(
    session_id: u32,
    file_path: String,
    source: String,
    version: i64,
    host: State<'_, EditorHost>,
) -> Result<LanguageFormattingResult, EditorError> {
    run_host_command(host, move |host| {
        host.sessions()
            .format_language_document(session_id, &file_path, &source, version)
    })
    .await
}

#[tauri::command]
pub(crate) async fn close_language_document(
    session_id: u32,
    file_path: String,
    host: State<'_, EditorHost>,
) -> Result<(), EditorError> {
    run_host_command(host, move |host| {
        host.sessions()
            .close_language_document(session_id, &file_path)
    })
    .await
}

#[tauri::command]
pub(crate) async fn function_document(
    session_id: u32,
    source: String,
    body: Option<String>,
    host: State<'_, EditorHost>,
) -> Result<FunctionDocumentState, EditorError> {
    run_host_command(host, move |host| {
        host.sessions()
            .function_document(session_id, &source, body.as_deref())
    })
    .await
}

#[tauri::command]
pub(crate) async fn write_source_text(
    session_id: u32,
    file_path: String,
    source: String,
    host: State<'_, EditorHost>,
) -> Result<ProjectBootstrap, EditorError> {
    run_host_command(host, move |host| {
        host.sessions()
            .write_source_text(session_id, &file_path, &source)
    })
    .await
}

File: studio/editors/cfd-editor/frontend/src/api/language.ts
import type { FlatDiagnostic } from '../bindings/FlatDiagnostic'
import type { FunctionDocumentState } from '../bindings/FunctionDocumentState'
import type { LanguageCompletion } from '../bindings/LanguageCompletion'
import type { LanguageDocumentState } from '../bindings/LanguageDocumentState'
import type { LanguageFormattingResult } from '../bindings/LanguageFormattingResult'
import type { LanguagePosition } from '../bindings/LanguagePosition'
import type { ProjectBootstrap } from '../bindings/ProjectBootstrap'
import { invokeCommand } from './tauriEnv'

export async function readSourceText(sessionId: number, filePath: string): Promise<string> {
  return invokeCommand<string>('read_source_text', { sessionId, filePath })
}

export async function highlightSourceSnapshot(sessionId: number, filePath: string, source: string): Promise<LanguageDocumentState> {
  return invokeCommand<LanguageDocumentState>('highlight_source_snapshot', { sessionId, filePath, source })
}

export async function syncLanguageDocument(
  sessionId: number,
  filePath: string,
  source: string,
  version: number,
): Promise<LanguageDocumentState> {
  return invokeCommand<LanguageDocumentState>('sync_language_document', {
    sessionId,
    filePath,
    source,
    version,
  })
}

export async function validateSourceText(
  sessionId: number,
  filePath: string,
  source: string,
): Promise<FlatDiagnostic[]> {
  return invokeCommand<FlatDiagnostic[]>('validate_source_text', { sessionId, filePath, source })
}

export async function formatLanguageDocument(
  sessionId: number,
  filePath: string,
  source: string,
  version: number,
): Promise<LanguageFormattingResult> {
  return invokeCommand<LanguageFormattingResult>('format_language_document', {
    sessionId,
    filePath,
    source,
    version,
  })
}

export async function completeLanguageDocument(
  sessionId: number,
  filePath: string,
  source: string,
  version: number,
  position: LanguagePosition,
): Promise<LanguageCompletion[]> {
  return invokeCommand<LanguageCompletion[]>('complete_language_document', {
    sessionId,
    filePath,
    source,
    version,
    position,
  })
}

export async function closeLanguageDocument(sessionId: number, filePath: string): Promise<void> {
  return invokeCommand('close_language_document', { sessionId, filePath })
}

export async function functionDocument(
  sessionId: number,
  source: string,
  body?: string,
): Promise<FunctionDocumentState> {
  return invokeCommand<FunctionDocumentState>('function_document', { sessionId, source, body })
}

export async function writeSourceText(
  sessionId: number,
  filePath: string,
  source: string,
  expectedSource: string,
): Promise<ProjectBootstrap> {
  return invokeCommand<ProjectBootstrap>('write_source_text', { sessionId, filePath, source, expectedSource })
}

